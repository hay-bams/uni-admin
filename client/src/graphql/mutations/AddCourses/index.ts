import { gql } from "@apollo/client";

export const ADD_COURSES = gql`
  mutation AddCourse($studentId: String!, $input: ID!) {
    addCourses(studentId: $studentId, input: $input) {
      id
    }
  }
`;