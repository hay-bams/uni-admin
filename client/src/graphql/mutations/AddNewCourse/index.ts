import { gql } from "@apollo/client";

export  const ADD_NEW_COURSE = gql`
  mutation AddNewCourse($input: NewCourseInput) {
    addNewCourse(input: $input) {
      id
    }
  }
`