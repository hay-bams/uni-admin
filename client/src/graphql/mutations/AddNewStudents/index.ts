import { gql } from "@apollo/client";

export  const ADD_NEW_STUDENT = gql`
  mutation AddNewStudent($input: NewStudentInput) {
    addNewStudent(input: $input) {
      id
    }
  }
`