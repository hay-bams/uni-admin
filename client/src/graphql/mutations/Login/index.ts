import { gql } from "@apollo/client";

export  const LOG_IN = gql`
  mutation Login($input: LoginInput) {
    login(input: $input) {
      id
      username
      token
      madeRequest
    }
  }
`