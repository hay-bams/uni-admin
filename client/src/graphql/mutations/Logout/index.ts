import { gql } from "@apollo/client";

export  const LOG_OUT = gql`
  mutation Logout {
    logout {
      id
      username
      madeRequest
    }
  }
`