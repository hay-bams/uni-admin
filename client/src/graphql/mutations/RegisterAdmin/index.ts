import { gql } from "@apollo/client";

export  const REGISTER_ADMIN = gql`
  mutation RegisterAdmin($input: RegisterInput) {
    register(input: $input) {
      id
      username
      madeRequest
    }
  }
`