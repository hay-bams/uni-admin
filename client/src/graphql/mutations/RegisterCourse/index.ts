import { gql } from "@apollo/client";

export const REGISTER_COURSE = gql`
  mutation RegisterCourse($studentId: String!, $input: ID!) {
    registerCourse(studentId: $studentId, input: $input) {
      id
    }
  }
`;