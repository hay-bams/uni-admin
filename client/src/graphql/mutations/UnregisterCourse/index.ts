import { gql } from "@apollo/client";

export const UNREGISTER_COURSE = gql`
  mutation UnregisterCourse($studentId: String!, $courseId: String!) {
    unregisterCourse(studentId: $studentId, courseId: $courseId) {
      id
    }
  }
`