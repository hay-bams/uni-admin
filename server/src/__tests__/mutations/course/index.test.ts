/* eslint-disable @typescript-eslint/no-non-null-assertion */
import jwt from 'jsonwebtoken';
import { COURSES_DATA, USER_DATA } from '../../test_data';
import { createTestServer } from '../../helpers';
import { TEST_ADD_NEW_COURSE, TEST_UPDATE_COURSE } from '../mutations';
import { ObjectID } from 'mongodb';

describe('Course Mutation', () => {
  test('should add a new course', async () => {
    const { mutate } = createTestServer({
      db: {
        courses: {
          insertOne: jest.fn(() => ({ ops: [COURSES_DATA[0]] })),
          findOne: jest.fn(() => null),
        },
        users: {
          findOne: jest.fn(() => USER_DATA[0]),
        },
      },
      req: {
        get: jest.fn(() => 'skskkdkdk'),
        signedCookies: {
          admin: USER_DATA[0]._id,
        },
      },
      res: {
        clearCookie: jest.fn(),
      },
    });

    jest.spyOn(jwt, 'verify').mockImplementation(
      jest.fn(() => ({
        admin: USER_DATA[0]._id,
      }))
    );

    const res = await mutate({
      mutation: TEST_ADD_NEW_COURSE,
      variables: {
        input: {
          name: COURSES_DATA[0].name,
          totalSeats: COURSES_DATA[0].totalSeats,
          status: COURSES_DATA[0].status,
        },
      },
    });

    expect(res.data.addNewCourse).toHaveProperty('id');
    expect(res.data.addNewCourse).toHaveProperty('name');
    expect(res.data.addNewCourse).toHaveProperty('totalSeats');
    expect(res).toMatchSnapshot();
  });

  test('should throw an error if course already exists', async () => {
    const { mutate } = createTestServer({
      db: {
        courses: {
          findOne: jest.fn(() => COURSES_DATA[0]),
        },
        users: {
          findOne: jest.fn(() => USER_DATA[0]),
        },
      },
      req: {
        get: jest.fn(() => 'skskkdkdk'),
        signedCookies: {
          admin: USER_DATA[0]._id,
        },
      },
      res: {
        clearCookie: jest.fn(),
      },
    });

    jest.spyOn(jwt, 'verify').mockImplementation(
      jest.fn(() => ({
        admin: USER_DATA[0]._id,
      }))
    );

    const res = await mutate({
      mutation: TEST_ADD_NEW_COURSE,
      variables: {
        input: {
          name: COURSES_DATA[0].name,
          totalSeats: COURSES_DATA[0].totalSeats,
          status: COURSES_DATA[0].status,
        },
      },
    });

    expect(res.errors?.length).toBeGreaterThan(0);
    expect(res.errors![0]).toHaveProperty('message');
    expect(res).toMatchSnapshot();
  });

  test('should update a course', async () => {
    const { mutate } = createTestServer({
      db: {
        courses: {
          findOneAndUpdate: jest.fn(() => ({ value: COURSES_DATA[0] })),
        },
        users: {
          findOne: jest.fn(() => USER_DATA[0]),
        },
      },
      req: {
        get: jest.fn(() => 'skskkdkdk'),
        signedCookies: {
          admin: USER_DATA[0]._id,
        },
      },
      res: {
        clearCookie: jest.fn(),
      },
    });

    jest.spyOn(jwt, 'verify').mockImplementation(
      jest.fn(() => ({
        admin: USER_DATA[0]._id,
      }))
    );


    const res = await mutate({
      mutation: TEST_UPDATE_COURSE,
      variables: {
        courseId: COURSES_DATA[0]._id.toString(),
        input: {
          name: COURSES_DATA[0].name,
          totalSeats: COURSES_DATA[0].totalSeats,
          status: COURSES_DATA[0].status,
        },
      },
    });

    expect(res.data.updateCourse).toHaveProperty('id');
    expect(res.data.updateCourse).toHaveProperty('name');
    expect(res.data.updateCourse).toHaveProperty('totalSeats');
    expect(res).toMatchSnapshot();
  });


  test('should throw an error if course to be updarted is not found', async () => {
    const { mutate } = createTestServer({
      db: {
        courses: {
          findOneAndUpdate: jest.fn(() => (null)),
        },
        users: {
          findOne: jest.fn(() => USER_DATA[0]),
        },
      },
      req: {
        get: jest.fn(() => 'skskkdkdk'),
        signedCookies: {
          admin: USER_DATA[0]._id,
        },
      },
      res: {
        clearCookie: jest.fn(),
      },
    });

    jest.spyOn(jwt, 'verify').mockImplementation(
      jest.fn(() => ({
        admin: USER_DATA[0]._id,
      }))
    );


    const res = await mutate({
      mutation: TEST_UPDATE_COURSE,
      variables: {
        courseId: COURSES_DATA[0]._id.toString(),
        input: {
          name: COURSES_DATA[0].name,
          totalSeats: COURSES_DATA[0].totalSeats,
          status: COURSES_DATA[0].status,
        },
      },
    });

    expect(res.errors?.length).toBeGreaterThan(0);
    expect(res.errors![0]).toHaveProperty('message');
    expect(res).toMatchSnapshot();
  });
});
