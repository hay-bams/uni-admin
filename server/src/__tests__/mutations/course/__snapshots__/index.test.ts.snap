// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course Mutation should add a new course 1`] = `
Object {
  "data": Object {
    "addNewCourse": Object {
      "id": "5fdab398791e134a5466566f",
      "name": "Aerospace Engineerng",
      "totalSeats": 25,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Course Mutation should throw an error if course already exists 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Something went wrong: Error: Course Already Exists],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
